2021-year-report.org

```diff
@@ -0,0 +1,438 @@
+
+#+BEGIN_SRC rust
+struct Foo(i64 bool);
+ 
+pub fn main() {
+   let a;
+   a = Foo{0: 123, 1: true};
+}
+#+END_SRC
+
+We hit this test case early on when trying different constructor variants for ADT's; TupleStructs are usually constructed using the regular Tuple syntax, though this test case demonstrates how the compiler implicitly creates implicit field names of the index. So it is almost an implementation detail that this works the way it does rather than being specified in the syntax of the language, which is pretty neat.
+
+see: https://godbolt.org/z/PEGEWae4f
+
+**** Qualified Paths
+
+As part of my work into traits, this includes qualified paths. I have been investigating some test cases around qualified paths, and this one had me confused: https://github.com/rust-lang/rust/blob/master/src/test/ui/qualified/qualified-path-params-2.rs.
+
+The associated path <S as Tr>::A on its own will resolve to unit-struct S so when the final segment of ::f<u8> i would have assumed this would have resolved to the type of the impl function f substituted with u8. I don't see why this is ambiguous. I have, however, received an explanation on the Rust Zulip server https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/Ambiguous.20associated.20types/near/251210283.
+
+This is interesting since associated types are not supported on impl blocks; the rustc compiler will enforce that all associated types must come from QualifiedPaths. This is to force the path to always be of the form <A as B>::C. When inherently associated types are stabilized, this will need some work. The GCC Rust implementation currently ICE's for this test case see https://github.com/Rust-GCC/gccrs/issues/843.
+
+See: [[https://doc.rust-lang.org/reference/paths.html#qualified-paths][QualifiedPathInType]]
+
+**** Never Type
+
+Rust has a never type which is pretty interesting. In c/c++, you can use the attribute no_return to signify that this function is never going to return. In Rust, they take this concept to its full with the never type, so for example, below: 
+
+#+BEGIN_SRC rust
+fn test() {
+   let a = return;
+   let b = a + 123;
+   a = 456;
+}
+#+END_SRC
+
+You can see 'a' is equal to the return expression, which will, in this context, never return. So even though everything after this, the first assignment will never get executed because of the return expression, Rust can fully type this function.  The functionality here is down to a key piece of the type system, which is more evident in closures than here. Still, in the above 'b = a + 123', the types here never type plus an integer inference variable, as there is no implementation of never type plus integer, here Rust actually returns another inference variable with a specified type-bound that requires the ability to add numbers together. This means the final assignment allows the type system to know that 'a' is an integer, albeit an uninitialized one, but this code is still safe because of the return expression.
```

Comment:
This is weirdly worded. 

So first of this compiles because of the assignment of `a = 456` at the end. With that the never type can be inferred to be `i32` (the fallback type for integers).

If the last line is removed this fails to compile because the never type falls back to `()` (which is still an open issue in rustc today).

The trait resolution doesn't help type checking here.

----

sidenote: the inline code snippets should be formatted with backticks `` ` ``

Reply:
grammerly has messed it up lol. This is org-mode not markdown btw.